onnxruntime/core/providers/cuda/tensor/identity_op.h

```diff
@@ -15,34 +16,57 @@ class IdentityOp final : public CudaKernel {
   }
 
   Status ComputeInternal(OpKernelContext* context) const override {
-    const Tensor* X = context->Input<Tensor>(0);
-    if (X == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
-    const TensorShape& shape = X->Shape();
-    Tensor* Y = context->Output(0, shape);
-    auto X_type = X->DataType();
+    auto X_ml_type = context->InputType(0);
+    if (DataTypeImpl::GetType<Tensor>() == X_ml_type) {
+      const Tensor* X = context->Input<Tensor>(0);
+      if (X == nullptr) return Status(common::ONNXRUNTIME, common::FAIL, "input count mismatch");
+      const TensorShape& shape = X->Shape();
+      Tensor* Y = context->Output(0, shape);
+      auto X_type = X->DataType();
 
-    const void* source = X->DataRaw(X_type);
-    void* target = Y->MutableDataRaw(X_type);
-    //If source and target pointers are not equal, we need to copy the data.
-    if (target != source) {
-      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(target, source, X->Shape().Size() * X->DataType()->Size(), cudaMemcpyDeviceToDevice, Stream()));
-    }
+      const void* source = X->DataRaw(X_type);
+      void* target = Y->MutableDataRaw(X_type);
+      //If source and target pointers are not equal, we need to copy the data.
+      if (target != source) {
```

Comment:
>  if (target != source) { [](http://example.com/codeflow?start=4&length=25)

The kernel has aliasing set, therefore, you will need to check if input and output are the same for sequence as well.

Reply:
Every tensor we copy to are created on the fly, there wouldn't be chance that source equals to target.